lib/burnside/intrinsics.cc

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "intrinsics.h"
+#include "builder.h"
+
+/// [Coding style](https://llvm.org/docs/CodingStandards.html)
+
+namespace Fortran::burnside {
+
+// Define a simple static runtime description that will be transformed into
+// IntrinsicImplementation when building the IntrinsicLibrary.
+namespace runtime {
+enum class Type { f32, f64 };
+struct StaticDescription {
+  const char *name;
+  const char *symbol;
+  Type resultType;
+  std::vector<Type> argumentTypes;
+};
+
+// TODO
+// This table should be generated in a clever ways and probably shared with
+// lib/evaluate intrinsic folding.
+static const StaticDescription llvm[] = {
+    {"abs", "llvm.fabs.f32", Type::f32, {Type::f32}},
+    {"abs", "llvm.fabs.f64", Type::f64, {Type::f64}},
+    {"acos", "acosf", Type::f32, {Type::f32}},
+    {"acos", "acos", Type::f64, {Type::f64}},
+    {"atan", "atan2f", Type::f32, {Type::f32, Type::f32}},
+    {"atan", "atan2", Type::f64, {Type::f64, Type::f64}},
+    {"sqrt", "llvm.sqrt.f32", Type::f32, {Type::f32}},
+    {"sqrt", "llvm.sqrt.f64", Type::f64, {Type::f64}},
+    {"cos", "llvm.cos.f32", Type::f32, {Type::f32}},
+    {"cos", "llvm.cos.f64", Type::f64, {Type::f64}},
+    {"sin", "llvm.sin.f32", Type::f32, {Type::f32}},
+    {"sin", "llvm.sin.f64", Type::f64, {Type::f64}},
+};
+
+// Conversion between types of the static representation and MLIR types.
+mlir::Type toMLIRType(Type t, mlir::MLIRContext &context) {
```

Comment:
Take a look at the code in fe-helper. All the type conversions should be there.

Reply:
Yeah, it is slightly different. It is not front-end dependent but more target dependent.
It takes `Type` from the enum just above. The thing is I am not sure how to statically "describe" the signature.

I cannot use f18/fir types either because the runtime signature are "fixed" i.e if I were to use real<4> in the description, compile time options might make acosf a f32 or an f64 depending on compile options. acosf is always f32. 
I am not sure how to directly create mlir type "statically"...

Alternative would be to directly generate (e.g. from a td file) the C++ code to dynamically build the `FunctionType` of functions instead of generating a table. But that would not be nice to read.
Or, I could store function pointers to the type generators inside the table (the `(mlir::Type)(*)(mlir::Context*)` functions).
something like:
```
namespace Type {
constexpr auto f32 = &mlir::FloatType::getF32;
}
static const StaticDescription llvm[] =
{"abs", "llvm.fabs.f32", Type::f32, {Type::f32}}
}
```
and then in `toMLIRFunctionType`

```
 for (auto typeBuilder : func.argumentTypes) {
    argMLIRTypes.push_back(typeBuilder(context));
  }
```

I am welcoming better suggestions.

Reply:
Kept as is for now.

Reply:
Kept as it is for now.